File: 0165-compare-version-numbers/0165-compare-version-numbers.java
class Solution {
    public int compareVersion(String v1, String v2) {
        int n1 = v1.length(), n2 = v2.length();
        int x1 = 0, x2 = 0;
        int i = 0, j = 0;
        while (i < n1 || j < n2) {
            while (i < n1 && v1.charAt(i) != '.') {
                x1 = 10 * x1 + (v1.charAt(i++) - '0');
            }
            while (j < n2 && v2.charAt(j) != '.') {
                x2 = 10 * x2 + (v2.charAt(j++) - '0');
            }
            if (x1 < x2) {
                return -1;
            } else if (x1 > x2) {
                return 1;
            }
            x1 = 0;
            x2 = 0;
            i++;
            j++;
        }
        return 0;
    }
}

File: 0165-compare-version-numbers/0165-compare-version-numbers.py
class Solution:
    def compareVersion(self, v1: str, v2: str) -> int:
        n1, n2 = len(v1), len(v2)
        x1, x2 = 0, 0
        i, j = 0, 0
        while i < n1 or j < n2:
            while i < n1 and v1[i] != '.':
                x1 = 10 * x1 + int(v1[i])
                i += 1
            while j < n2 and v2[j] != '.':
                x2 = 10 * x2 + int(v2[j])
                j += 1
            if x1 < x2:
                return -1
            elif x1 > x2:
                return 1
            x1, x2 = 0, 0
            i += 1
            j += 1
        return 0